raft.go

```diff
@@ -254,6 +254,23 @@ type Config struct {
 	// logical clock from assigning the timestamp and then forwarding the data
 	// to the leader.
 	DisableProposalForwarding bool
+
+	// DisableConfChangeValidation turns off propose-time verification of
+	// configuration changes against the currently active configuration of the
+	// raft instance. These checks are generally sensible (cannot leave a joint
+	// config unless in a joint config, et cetera) but they have false positives
+	// because the active configuration may not be the most recent
+	// configuration. This is because configurations are activated during log
+	// application, and even the leader can trail log application by an
+	// unbounded number of entries.
+	// Symmetrically, the mechanism has false negatives - because the check may
+	// not run against the "actual" config that will be the predecessor of the
+	// newly proposed one, the check may pass but the new config may be invalid
+	// when it is being applied. In other words, the checks are best-effort.
+	//
+	// Users should *not* use this option unless they have a reliable mechanism
+	// (above raft) that serializes and verifies configuration changes.
+	DisableConfChangeValidation bool
```

Comment:
I think the exposed configuration item is confusing to users.  If there are any issues on the conf change validation or conf change itself, then we should only fix the raft side instead of delegating part of the responsibility/fix to upper applications. Providing more flexibility here (exposing the conf item) also provides a dangerous hole for users to make mistake (e.g. turning on the item: setting it to `true` wrongly).

Usually users (upper applications) completely depends on raft to reach an agreement on the conf change. They shouldn't worry about anything on conf change at all. 

Can you just workaround this issue on application side for now, and In the meanwhile, reconsider how to improve the conf change in raft in the long-term?

Reply:
> Can you just workaround this issue on application side for now, and In the meanwhile, reconsider how to improve the conf change in raft in the long-term?

That was what I tried at first, unfortunately it was not successful and convinced me that it is not the right approach here. The invariant we want in CockroachDB is that if you make a proposal and hand it to raft often enough, it will eventually come up for application (where CockroachDB determines whether it will have an effect or apply as a no-op, subject to our replay protection mechanism, lease system, etc). This check breaks this invariant, and since the check is unstable (produces different results at different times) there is the scenario in which we cannot tell whether the proposal made it into the log, which means we can't release certain locks until we see it in application. It's a bit complicated to explain, but essentially working around it in CRDB is something I will rule out on account of the undue complexity it imposes.

> If there are any issues on the conf change validation or conf change itself, then we should only fix the raft side instead of delegating part of the responsibility/fix to upper applications

I would describe what I am doing here as a tactical fix. The incorrect check in raft is a problem, and I agree with you that it should be addressed (which is tracked in https://github.com/etcd-io/raft/issues/80). However, I expect that to be a nontrivial change, and there should be a way to opt out of the problematic behavior until then - which is what this patch achieves.

> Usually users (upper applications) completely depends on raft to reach an agreement on the conf change. They shouldn't worry about anything on conf change at all.

This might be true in most simple systems, but it is not true in CockroachDB. We currently avoid drift between the state machine's tracked config and the raft config, and we do so by updating the tracked config through raft conf changes. So it's not acceptable to us for raft to drop configuration changes that would be valid. And at the same time, it means that we have necessarily already validated the configuration change before it happens and establish mutual exclusion of subsequent conf changes.
I can't rule out that some of these boundary conditions will change in the future, but I don't see it as an anti pattern regardless. The raft library should be useful for advanced use as well as simple use cases.

Reply:
thx for the clarification. At the minimum, I think we need to clearly add a comment to mention that `DisableConfChangeValidation` is just a temporary solution for https://github.com/etcd-io/raft/issues/80, may be removed after the issue has been resolved in raft.

Reply:
Good point - added the comment. PTAL.